nat-lab/tests/test_direct_connection.py

```diff
@@ -189,27 +191,30 @@ async def test_direct_working_paths(
             for client, node in itertools.product(env.clients, env.nodes)
             if not client.is_node(node)
         ])
-
-        direct_events = await asyncio.gather(*[
-            exit_stack.enter_async_context(
-                run_async_context(
-                    client.wait_for_event_peer(
-                        node.public_key, [NodeState.CONNECTED], [PathType.DIRECT]
-                    )
-                )
-            )
-            for client, node in itertools.product(env.clients, env.nodes)
-            if not client.is_node(node)
-        ])
-
         async with _disable_direct_connection(env, reflexive_ips):
             await asyncio.gather(*relay_events)
             await ping_between_all_nodes(env)
 
+            # Start collecting direct events, before drop of blocking
+            print(datetime.now(), "Reconnect to direct connections")
+            direct_events = await asyncio.gather(*[
+                exit_stack.enter_async_context(
+                    run_async_context(
+                        client.wait_for_event_peer(
+                            node.public_key, [NodeState.CONNECTED], [PathType.DIRECT]
+                        )
+                    )
+                )
+                for client, node in itertools.product(env.clients, env.nodes)
+                if not client.is_node(node)
+            ])
```

Comment:
Isn't this in the wrong scope now? It looks from the indentation that it's part of the `_disable_direct_connection` blocks whereas the `await gather` call for it is outside of that scope. At least that's what it looks like

Reply:
Yeh this is the anoying part, its future in futures:
```
async def show():
    async def fake_wait():
        print("done")
        return "done"
        
    async with AsyncExitStack() as exit_stack:
        x = exit_stack.enter_async_context(
            run_async_context(
                fake_wait()
            )
        )
        print("x", x)
        
        y = await x
        print("y", y)

        z = await y
        print("z", z)

asyncio.run(show())
```

outputs:
```
x <coroutine object AsyncExitStack.enter_async_context at 0x77509351c040>
y <Task pending name='Task-2' coro=<show.<locals>.fake_wait() running at /home/etg/etg/src/telio-build/libtelio/nat-lab/./tests/test_direct_connection.py:702>>
done
z done
```

Reply:
basicaly enter_async_context is also a future, so needs to be awaited, then returns a wrapped coroutine